unit-tests-java8/src/test/java/org/eclipse/collections/test/RichIterableTestCase.java

```diff
@@ -837,6 +852,65 @@ default void RichIterable_iterationOrder()
         });
         assertEquals(expectedIterationOrder, sumOfLongIterationOrder);
 
+        /*
+         * TODO: Fix sumByPrimitive methods for bags, to only iterate once per item, not per occurrence.
+        MutableCollection<Integer> sumByDoubleIterationOrder1 = this.newMutableForFilter();
+        MutableCollection<Integer> sumByDoubleIterationOrder2 = this.newMutableForFilter();
+        this.getInstanceUnderTest().sumByDouble(
+                each -> {
+                    sumByDoubleIterationOrder1.add(each);
+                    return each;
+                },
+                each -> {
+                    sumByDoubleIterationOrder2.add(each);
+                    return 0.0;
+                });
+        assertEquals(expectedIterationOrder, sumByDoubleIterationOrder1);
+        assertEquals(expectedIterationOrder, sumByDoubleIterationOrder2);
+
+        MutableCollection<Integer> sumByFloatIterationOrder1 = this.newMutableForFilter();
+        MutableCollection<Integer> sumByFloatIterationOrder2 = this.newMutableForFilter();
+        this.getInstanceUnderTest().sumByFloat(
+                each -> {
+                    sumByFloatIterationOrder1.add(each);
+                    return each;
+                },
+                each -> {
+                    sumByFloatIterationOrder2.add(each);
+                    return 0.0f;
+                });
+        assertEquals(expectedIterationOrder, sumByFloatIterationOrder1);
+        assertEquals(expectedIterationOrder, sumByFloatIterationOrder2);
+
+        MutableCollection<Integer> sumByIntIterationOrder1 = this.newMutableForFilter();
+        MutableCollection<Integer> sumByIntIterationOrder2 = this.newMutableForFilter();
+        this.getInstanceUnderTest().sumByInt(
+                each -> {
+                    sumByIntIterationOrder1.add(each);
+                    return each;
+                },
+                each -> {
+                    sumByIntIterationOrder2.add(each);
+                    return 0;
+                });
+        assertEquals(expectedIterationOrder, sumByIntIterationOrder1);
+        assertEquals(expectedIterationOrder, sumByIntIterationOrder2);
+
+        MutableCollection<Integer> sumByLongIterationOrder1 = this.newMutableForFilter();
+        MutableCollection<Integer> sumByLongIterationOrder2 = this.newMutableForFilter();
+        this.getInstanceUnderTest().sumByLong(
+                each -> {
+                    sumByLongIterationOrder1.add(each);
+                    return each;
+                },
+                each -> {
+                    sumByLongIterationOrder2.add(each);
+                    return 0L;
+                });
+        assertEquals(expectedIterationOrder, sumByLongIterationOrder1);
+        assertEquals(expectedIterationOrder, sumByLongIterationOrder2);
+        */
```

Comment:
While writing tests to cover OrderedMapAdapter, I found some unrelated problems. Here's the first one. The sumBy* methods were essentially uncovered by tests. Now that they are covered, the tests fail for all bag implementations, since the key function is being invoked once per occurrence rather than once per unique item.

Reply:
Sorry! I don't quite follow this. If we are doing a `sumByPrimitive()` on a List, we will evaluate the function over each element, even duplicates and do an update to the value as in org.eclipse.collections.impl.block.factory.PrimitiveFunctions#sumByDoubleFunction. For Bags also it should be the same no?
Let's consider this:
```
Bag<Integer> bag = Bags.mutable.with(1, 2, 2, 2, 3, 3, 3, 3);
MutableObjectDoubleMap result = bag.sumByDouble(each -> each, each -> 1);
MutableObjectDoubleMap expected = ObjectDoubleMaps.mutable.empty();

// Should the expected be this:
expected.put(1, 1);
expected.put(2, 3);
expected.put(3, 4);
// Or should the expected be this:
expected.put(1, 1);
expected.put(2, 1);
expected.put(3, 1);
```
I think it should be the former. But I could be wrong. What do you think @donraab , @motlin ?

Reply:
This isn't about the value of the expect result. It's about how many times the function gets called.

Reply:
Ah! If I understand correctly, in the implementation, we would should call the `groupBy` function and `DoubleFunction` just once, but the result of addition has to be compounded over the number of occurrences for each distinct item. Or I am missing something?

Reply:
Yes. Look at AbstractBag. Almost every method delegates to forEachWithOccurrences instead of forEach. That includes sumOf*. sumBy* need similar implementations in AbstractMutableBagIterable or below.

Reply:
Looks like the aggregateBy methods have the same problem.

Reply:
Sounds like a reasonable performance fix.

Reply:
I think we can only do the fix in major version. Seems like we have a lot of changes for 10.0.0 (October '18). To give us time we can close 9.2.0 in April and open 10.0.0 and start fixing all these issues. Thoughts?

Reply:
Raised #448 and #449